test/tests/functional/pbs_maintenance_reservations.py

```diff
@@ -0,0 +1,192 @@
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2'),
+                    'Resource_List.select':
+                    'host=%s:ncpus=4' % self.mom.shortname,
+                    'Resource_List.place': 'exclhost',
+                    'resv_nodes': '(vn[0]:ncpus=2)+(vn[1]:ncpus=2)'}
+        self.server.expect(RESV, exp_attr, id=rid)
+
+    def test_maintenance_degrade_reservation(self):
+        """
+        Test if the reservation is degraded by overlapping
+        maintenance reservation.
+        """
+        now = int(time.time())
+
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'managers': '%s@*' % TEST_USER})
+
+        a1 = {'Resource_List.select': 'host=%s' % self.mom.shortname,
+              'reserve_start': now + 3600,
+              'reserve_end': now + 3605}
+        r1 = Reservation(TEST_USER, attrs=a1)
+        rid1 = self.server.submit(r1)
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        self.server.expect(RESV, exp_attr, id=rid1)
+
+        a2 = {'reserve_start': now + 3602,
+              'reserve_end': now + 3604,
+              '--hosts': self.mom.shortname}
+        r2 = Reservation(TEST_USER, attrs=a2)
+        self.server.submit(r2)
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_DEGRADED|10'),
+                    'reserve_substate': 12}
+        self.server.expect(RESV, exp_attr, id=rid1)
+
+    def test_maintenance_reconfirm_resv(self):
+        """
+        Test if the overlapping reservation is reconfirmed. Two moms
+        (-p "servers=M1,moms=M1:M2") are needed for this test.
+        """
+        now = int(time.time())
+
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'managers': '%s@*' % TEST_USER})
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'scheduler_iteration': 3})
+
+        if len(self.moms) != 2:
+            cmt = "need 2 mom hosts: -p servers=<m1>,moms=<m1>:<m2>"
+            self.skip_test(reason=cmt)
+
+        self.momA = self.moms.values()[0]
+        self.momB = self.moms.values()[1]
+
+        a1 = {'Resource_List.select': '1:ncpus=1',
+              'reserve_start': now + 3600,
+              'reserve_end': now + 3605}
+        r1 = Reservation(TEST_USER, attrs=a1)
+        rid1 = self.server.submit(r1)
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2'),
+                    'resv_nodes': '(%s:ncpus=1)' % self.momA.shortname}
+        self.server.expect(RESV, exp_attr, id=rid1)
+
+        a2 = {'reserve_start': now + 3602,
+              'reserve_end': now + 3604,
+              '--hosts': self.momA.shortname}
+        r2 = Reservation(TEST_USER, attrs=a2)
+        self.server.submit(r2)
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2'),
+                    'resv_nodes': '(%s:ncpus=1)' % self.momB.shortname}
+        self.server.expect(RESV, exp_attr, id=rid1)
```

Comment:
Consider more tests (either some or all of these ideas).  Your maintenance reservation starts after the overlapping reservation and end before it.  Write one where the maintenance reservation starts before the overlapping reservation and ends during it.  Another where the maintenance reservation starts during the overlapping reservation and ends after it.  The third is where the maintenance reservation starts before and ends after the overlapping reservation.   Write a test where the two reservations overlap in nodes but not time and make sure the normal reservation is not marked degraded.  Write an end to end test.  Submit a maintenance reservation, see it overlap with another reservation, and make sure they work properly when they are running.  Write one where you see an overlapping reservation get reconfirmed after it is degraded.

Reply:
Thank you for suggestions. Several new tests have been added. I also added tests on top of suggested.